---
title: Walkthrough data
summary: Example data for walkthrough and testing
permalink: walkthrough_data.html
datatable: true
---
{% assign walkthrough_folder = "walkthrough/data/" %}

This page describes the dataset that is used in the 
[Walkthrough](walkthrough.html) for testing _ReadTools_ binaries.
A checksum for available files can be found 
[here]({{walkthrough_folder}}/checksum).

The data is a modified subset of reads from 
[SRR1931701](https://www.ncbi.nlm.nih.gov/sra/?term=SRR1931701).
Concretely, the first 100 read-pairs were downloaded. Out of this 100 
read-pairs, 3 were modified by masking read-ends (MASKED reads). 
Additionally,  the quality of 1 was changed (LOWQUAL).
These 103 read-pairs were encoded in the different formats 
mentioned below ([Reads data summary](#reads-data-summary)),
with the following additions/changes:

* Single end: only the first read of each pair was kept and marked as 
  unpaired following the specifications of each format.
* Quality encoding: a subset of files were duplicated and converted
  to the Illumina encoding for testing quality detection/correction.
* Read name: a subset of files were duplicated and read names were 
  changed according to different conventions (Illumina-legacy and 
  Casava) for testing lossless conversion and barcode handling.
* Barcodes: single or dual barcoding indexes based on a default set of
  barcodes ([Other data](#other-data)) were added to the files for 
  testing lossless conversion and de-multiplexing.

For both unmapped and mapped reads, the sort order is _unsorted_ but 
they are grouped by queryname (not reflected in the header). Mapped
reads were generated by [bwa]({{site.data.software.bwa}}) mem 
version 0.7.12 and default parameters.

## Reads data summary

A brief description for each dataset is included in the following table.

{% include note.html 
content="Click on the file name to download the file." %}
{% include tip.html
content="If SAM formatted reads are not available in BAM/CRAM format, 
please use [samtools](http://www.htslib.org/doc/samtools.html) view for conversion." %}

<!-- Note: Try to keep the columns to 3 or 4 columns only.
If you add 5+ columns, your table may create horizontal scrolling with 
the theme. Additionally, keep the column heading titles short. -->
<table class="display">
    <thead>
        <tr class="header">
            <th style="text-align:left" style="vertical-align:middle">File(s)</th>
            <th style="text-align:center" style="vertical-align:middle">Format</th>
            <th style="text-align:center" style="vertical-align:middle">Pair-end</th>
            <th style="text-align:center" style="vertical-align:middle">Quality<br>encoding</th>
            <th style="text-align:left" style="vertical-align:middle">Read<br>name</th>
            <th style="text-align:left" style="vertical-align:middle">Indexing</th>
            <th style="text-align:left" style="vertical-align:middle">Barcode<br>encoding</th>
        </tr>
    </thead>
    <tbody>
        {% for example_data in site.data.walkthrough_data %}
        <tr>
            <td style="text-align:left" style="vertical-align:middle">
            {% for file in example_data.files %}
                <a href="{{walkthrough_folder}}{{file}}">{{file}}</a>
                {% unless forloop.last %}<br>{% endunless %}
            {% endfor %}
                
            </td>
            <td style="text-align:center" style="vertical-align:middle">{{example_data.format}}</td>
            <td style="text-align:center" style="vertical-align:middle">{{example_data.paired}}</td>
            <td style="text-align:center" style="vertical-align:middle">{{example_data.quality}}</td>
            <td style="text-align:left" style="vertical-align:middle">{{example_data.readname}}</td>
            <td style="text-align:left" style="vertical-align:middle">{{example_data.indexing}}</td>
            <td style="text-align:left" style="vertical-align:middle">{{example_data.barcode}}</td>
        </tr>
        {% endfor %}
    </tbody>

</table>

## Other data

Additional required data beside of read sources:

* [2L.fragment.fa]({{walkthrough_folder}}2L.fragment.fa): 
  FASTA file for use with mapped reads and CRAM sources. The folder 
  also includes the following files needed for its use in ReadTools:
  - [2L.fragment.fa.fai]({{walkthrough_folder}}2L.fragment.fa.fai): index file
  - [2L.fragment.dict]({{walkthrough_folder}}2L.fragment.dict): dictionary (SAM header)
* [dual.barcodes]({{walkthrough_folder}}dual.barcodes): barcode file 
  for samples with dual-indexing.
* [single.barcodes]({{walkthrough_folder}}single.barcodes): barcode 
  file for samples with single-indexing.
